cmd/restic/cmd_mount.go

```diff
@@ -0,0 +1,301 @@
+	}
+
+	repo, err := cmd.global.OpenRepository()
+	if err != nil {
+		return err
+	}
+
+	err = repo.LoadIndex()
+	if err != nil {
+		return err
+	}
+
+	mountpoint := args[0]
+	if _, err := os.Stat(mountpoint); err != nil {
+		if os.IsNotExist(err) {
+			cmd.global.Verbosef("Mountpoint %s doesn't exist, creating it\n", mountpoint)
+			err = os.Mkdir(mountpoint, os.ModeDir|0700)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	c, err := fuse.Mount(
+		mountpoint,
+		fuse.ReadOnly(),
+		fuse.FSName("restic"),
+	)
+	if err != nil {
+		return err
+	}
+
+	root := fs.Tree{}
+	root.Add("snapshots", &snapshots{
+		repo:           repo,
+		knownSnapshots: make(map[string]*restic.Snapshot),
+	})
+
+	cmd.global.Printf("Now serving %s under %s\n", repo.Backend().Location(), mountpoint)
+	cmd.global.Printf("Don't forget to umount after quitting !\n")
+
+	cmd.ready <- struct{}{}
+
+	err = fs.Serve(c, &root)
+	if err != nil {
+		return err
+	}
+
+	<-c.Ready
+	return c.MountError
+}
+
+// These lines statically ensure that a *snapshots implement the given
+// interfaces; a misplaced refactoring of the implementation that breaks
+// the interface will be catched by the compiler
+var _ = fs.HandleReadDirAller(&snapshots{})
+var _ = fs.NodeStringLookuper(&snapshots{})
+
+type snapshots struct {
+	repo *repository.Repository
+
+	// snapshot timestamp -> snapshot
+	knownSnapshots map[string]*restic.Snapshot
+}
+
+func (sn *snapshots) Attr(ctx context.Context, a *fuse.Attr) error {
+	a.Inode = 0
+	a.Mode = os.ModeDir | 0555
+	return nil
+}
+
+func (sn *snapshots) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
+	ret := make([]fuse.Dirent, 0)
+	for id := range sn.repo.List(backend.Snapshot, ctx.Done()) {
+		snapshot, err := restic.LoadSnapshot(sn.repo, id)
+		if err != nil {
+			return nil, err
+		}
+		sn.knownSnapshots[snapshot.Time.Format(time.RFC3339)] = snapshot
+		ret = append(ret, fuse.Dirent{
+			Inode: binary.BigEndian.Uint64(id[:8]),
+			Type:  fuse.DT_Dir,
+			Name:  snapshot.Time.Format(time.RFC3339),
+		})
+	}
+
+	return ret, nil
+}
+
+func (sn *snapshots) Lookup(ctx context.Context, name string) (fs.Node, error) {
+	if _, ok := sn.knownSnapshots[name]; !ok {
+		// At least this snapshot is not cached. We use this opportunity to
+		// load all missing snapshots
+		for id := range sn.repo.List(backend.Snapshot, ctx.Done()) {
+			snapshot, err := restic.LoadSnapshot(sn.repo, id)
+			if err != nil {
+				return nil, err
+			}
+			snapshotName := snapshot.Time.Format(time.RFC3339)
+			sn.knownSnapshots[snapshotName] = snapshot
+			break
+		}
+	}
+
+	snapshot := sn.knownSnapshots[name]
+	tree, err := restic.LoadTree(sn.repo, snapshot.Tree)
+	if err != nil {
+		return nil, err
+	}
+	return &dir{
+		repo: sn.repo,
+		tree: tree,
+	}, nil
+	return nil, fuse.ENOENT
+}
+
+// Statically ensure that *dir implement those interface
+var _ = fs.HandleReadDirAller(&dir{})
+var _ = fs.NodeStringLookuper(&dir{})
+
+type dir struct {
```

Comment:
Hmm I think this should be a separate file too (and it's methods). Not a fan of having the cmd stuff and two separate types all in one big 300 line file.